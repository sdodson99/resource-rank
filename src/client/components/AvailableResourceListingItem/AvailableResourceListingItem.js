import React from 'react';
import PropTypes from 'prop-types';
import styles from './AvailableResourceListingItem.module.css';
import ListingItem from '../ListingItem/ListingItem';
import { Add } from '@material-ui/icons';
import VerifiedIcon from '../VerifiedIcon/VerifiedIcon';

const AvailableResourceListingItem = ({ resource, onAdd }) => {
  const { name, verified, alreadyAdded, hasAddError } = resource;

  const onAddClick = () => onAdd && onAdd(resource);

  const calculateNameDisplayClassName = () => {
    if (alreadyAdded) {
      return 'line-through';
    }

    return '';
  };

  return (
    <div
      className={styles.AvailableResourceListingItem}
      data-testid="AvailableResourceListingItem"
    >
      <ListingItem>
        <div className="text-center sm:text-left sm:flex sm:justify-between sm:items-center">
          <div>
            <div className="flex items-center justify-center sm:justify-start">
              <div className={calculateNameDisplayClassName()}>{name}</div>
              {verified && (
                <div className="ml-2">
                  <VerifiedIcon />
                </div>
              )}
            </div>
            {alreadyAdded && (
              <div className="error-text italic text-xs">Already added</div>
            )}
            {hasAddError && (
              <div className="error-text text-xs">
                Failed to add topic resource.
              </div>
            )}
          </div>

          <div className="mt-5 sm:mt-0 sm:ml-5">
            <button
              className="btn btn-primary"
              onClick={onAddClick}
              disabled={alreadyAdded}
              data-testid="AddResourceButton"
            >
              <Add fill="white" />
            </button>
          </div>
        </div>
      </ListingItem>
    </div>
  );
};

AvailableResourceListingItem.propTypes = {
  resource: PropTypes.shape({
    name: PropTypes.string,
    verified: PropTypes.bool,
    alreadyAdded: PropTypes.bool,
    hasAddError: PropTypes.bool,
  }),
  onAdd: PropTypes.func,
};

AvailableResourceListingItem.defaultProps = {
  resource: {
    name: '',
  },
};

export default AvailableResourceListingItem;
