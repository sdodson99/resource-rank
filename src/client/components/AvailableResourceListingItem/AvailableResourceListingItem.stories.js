import React from 'react';
import AvailableResourceListingItem from './AvailableResourceListingItem';

export default {
  component: AvailableResourceListingItem,
  title: 'Components/AvailableResourceListingItem',
};

const Template = (args) => <AvailableResourceListingItem {...args} />;

export const Default = Template.bind({});
Default.args = {
  resource: {
    name: 'Resource 1',
  },
};

export const Verified = Template.bind({});
Verified.args = {
  resource: {
    name: 'Resource 1',
    verified: true,
  },
};

export const AlreadyAdded = Template.bind({});
AlreadyAdded.args = {
  resource: {
    name: 'Resource 1',
    alreadyAdded: true,
  },
};

export const VerifiedAlreadyAdded = Template.bind({});
VerifiedAlreadyAdded.args = {
  resource: {
    name: 'Resource 1',
    alreadyAdded: true,
    verified: true,
  },
};

export const Error = Template.bind({});
Error.args = {
  resource: {
    name: 'Resource 1',
    hasAddError: true,
  },
};
