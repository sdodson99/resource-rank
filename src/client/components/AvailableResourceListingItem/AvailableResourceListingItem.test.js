import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import AvailableResourceListingItem from './AvailableResourceListingItem';
import renderer from 'react-test-renderer';

describe('<AvailableResourceListingItem />', () => {
  let resource;

  beforeEach(() => {
    resource = {
      id: '123',
    };
  });

  it('should mount', () => {
    render(<AvailableResourceListingItem />);

    const availableResourceListingItem = screen.getByTestId(
      'AvailableResourceListingItem'
    );

    expect(availableResourceListingItem).toBeInTheDocument();
  });

  it('should render correctly', () => {
    const tree = renderer
      .create(<AvailableResourceListingItem resource={resource} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when resource verified', () => {
    resource.verified = true;

    const tree = renderer
      .create(<AvailableResourceListingItem resource={resource} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call onAdd when added', () => {
    const onAdd = jest.fn();
    render(<AvailableResourceListingItem resource={resource} onAdd={onAdd} />);
    const addResourceButton = screen.getByTestId('AddResourceButton');

    fireEvent.click(addResourceButton);

    expect(onAdd).toBeCalledWith(resource);
  });

  it('should render correctly when resource has add error', () => {
    resource.hasAddError = true;

    const tree = renderer
      .create(<AvailableResourceListingItem resource={resource} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when resource already added', () => {
    resource.alreadyAdded = true;

    const tree = renderer
      .create(<AvailableResourceListingItem resource={resource} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
