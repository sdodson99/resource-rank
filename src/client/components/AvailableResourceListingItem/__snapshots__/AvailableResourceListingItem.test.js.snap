// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvailableResourceListingItem /> should render correctly 1`] = `
<div
  className="AvailableResourceListingItem"
  data-testid="AvailableResourceListingItem"
>
  <div
    className="ListingItem"
    data-testid="ListingItem"
  >
    <div
      className="text-center sm:text-left sm:flex sm:justify-between sm:items-center"
    >
      <div>
        <div
          className="flex items-center justify-center sm:justify-start"
        >
          <div
            className=""
          />
        </div>
      </div>
      <div
        className="mt-5 sm:mt-0 sm:ml-5"
      >
        <button
          className="btn btn-primary"
          data-testid="AddResourceButton"
          onClick={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            fill="white"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableResourceListingItem /> should render correctly when adding 1`] = `
<div
  className="AvailableResourceListingItem"
  data-testid="AvailableResourceListingItem"
>
  <div
    className="ListingItem"
    data-testid="ListingItem"
  >
    <div
      className="text-center sm:text-left sm:flex sm:justify-between sm:items-center"
    >
      <div>
        <div
          className="flex items-center justify-center sm:justify-start"
        >
          <div
            className=""
          />
        </div>
      </div>
      <div
        className="mt-5 sm:mt-0 sm:ml-5"
      >
        <button
          className="btn btn-primary"
          data-testid="AddResourceButton"
          disabled={true}
          onClick={[Function]}
        >
          <div>
            <div
              className="LoadingSpinner"
              data-testid="LoadingSpinner"
            >
              <svg
                fill="none"
                height={20}
                viewBox="0 0 100 100"
                width={20}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M95 50C95 59.6293 91.9111 69.0047 86.1874 76.7483C80.4636 84.4919 72.4069 90.1952 63.2012 93.0201C53.9956 95.845 44.1265 95.6424 35.0445 92.4421C25.9625 89.2419 18.1466 83.2127 12.7454 75.2408C7.34427 67.2689 4.64275 57.7746 5.03791 48.1534C5.43306 38.5321 8.90405 29.2914 14.9407 21.7892C20.9774 14.287 29.2615 8.91914 38.5753 6.47442C47.8891 4.0297 57.7416 4.6371 66.6846 8.20736"
                  stroke="white"
                  strokeWidth="10"
                />
              </svg>
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableResourceListingItem /> should render correctly when can not add resource 1`] = `
<div
  className="AvailableResourceListingItem"
  data-testid="AvailableResourceListingItem"
>
  <div
    className="ListingItem"
    data-testid="ListingItem"
  >
    <div
      className="text-center sm:text-left sm:flex sm:justify-between sm:items-center"
    >
      <div>
        <div
          className="flex items-center justify-center sm:justify-start"
        >
          <div
            className=""
          />
        </div>
      </div>
      <div
        className="mt-5 sm:mt-0 sm:ml-5"
      >
        <button
          className="btn btn-primary"
          data-testid="AddResourceButton"
          disabled={true}
          onClick={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            fill="white"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableResourceListingItem /> should render correctly when resource already added 1`] = `
<div
  className="AvailableResourceListingItem"
  data-testid="AvailableResourceListingItem"
>
  <div
    className="ListingItem"
    data-testid="ListingItem"
  >
    <div
      className="text-center sm:text-left sm:flex sm:justify-between sm:items-center"
    >
      <div>
        <div
          className="flex items-center justify-center sm:justify-start"
        >
          <div
            className="line-through"
          />
        </div>
        <div
          className="error-text italic text-xs"
        >
          Already added
        </div>
      </div>
      <div
        className="mt-5 sm:mt-0 sm:ml-5"
      >
        <button
          className="btn btn-primary"
          data-testid="AddResourceButton"
          disabled={true}
          onClick={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            fill="white"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableResourceListingItem /> should render correctly when resource has add error 1`] = `
<div
  className="AvailableResourceListingItem"
  data-testid="AvailableResourceListingItem"
>
  <div
    className="ListingItem"
    data-testid="ListingItem"
  >
    <div
      className="text-center sm:text-left sm:flex sm:justify-between sm:items-center"
    >
      <div>
        <div
          className="flex items-center justify-center sm:justify-start"
        >
          <div
            className=""
          />
        </div>
        <div
          className="error-text text-xs"
        >
          Failed to add topic resource.
        </div>
      </div>
      <div
        className="mt-5 sm:mt-0 sm:ml-5"
      >
        <button
          className="btn btn-primary"
          data-testid="AddResourceButton"
          onClick={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            fill="white"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<AvailableResourceListingItem /> should render correctly when resource verified 1`] = `
<div
  className="AvailableResourceListingItem"
  data-testid="AvailableResourceListingItem"
>
  <div
    className="ListingItem"
    data-testid="ListingItem"
  >
    <div
      className="text-center sm:text-left sm:flex sm:justify-between sm:items-center"
    >
      <div>
        <div
          className="flex items-center justify-center sm:justify-start"
        >
          <div
            className=""
          />
          <div
            className="ml-2"
          >
            <div
              className="VerifiedIcon"
              data-testid="VerifiedIcon"
            >
              <svg
                aria-describedby={null}
                aria-hidden={true}
                aria-label="verified"
                className="MuiSvgIcon-root"
                color="#6f99f2"
                focusable="false"
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseLeave={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "fontSize": 20,
                  }
                }
                title="Verified"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="mt-5 sm:mt-0 sm:ml-5"
      >
        <button
          className="btn btn-primary"
          data-testid="AddResourceButton"
          onClick={[Function]}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            fill="white"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
